server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -469,38 +454,112 @@ protected enum SyntheticSourceMode {
-     * If mapper does not support synthetic source, it is generated using generic implementation
+     * Returns synthetic field loader for the mapper.
+     * If mapper does not support synthetic source, it is handled using generic implementation
      * in {@link DocumentParser#parseObjectOrField} and {@link ObjectMapper#syntheticFieldLoader()}.
+     * <br>
+     *
+     * This method is final in order to support common use cases like fallback synthetic source.
+     * Mappers that need custom support of synthetic source should override {@link #syntheticSourceSupport()}.
      *
      * @return implementation of {@link SourceLoader.SyntheticFieldLoader}
      */
     @Override
-    public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
-        // If mapper supports synthetic source natively, it overrides this method,
-        // so we won't see those here.
+    public final SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
+        if (hasScript()) {
+            return SourceLoader.SyntheticFieldLoader.NOTHING;
+        }
+
         if (syntheticSourceMode() == SyntheticSourceMode.FALLBACK) {
-            if (builderParams.copyTo.copyToFields().isEmpty() != true) {
-                throw new IllegalArgumentException(
-                    "field [" + fullPath() + "] of type [" + typeName() + "] doesn't support synthetic source because it declares copy_to"
-                );
-            }
-            // Nothing because it is handled at `ObjectMapper` level.
+            // Nothing because it is handled at `DocumentParser/ObjectMapper` level.
             return SourceLoader.SyntheticFieldLoader.NOTHING;
         }
 
-        return super.syntheticFieldLoader();
+        return syntheticSourceSupport().loader();
+    }
+
+    /**
+     * <p>
+     * Returns implementation of synthetic source support for the mapper.
+     * <br>
+     * By default (meaning {@link SyntheticSourceSupport.Fallback}),
+     * an exact full copy of parsed field value is stored separately
+     * and used for synthetic source.
+     * </p>
+     * <p>
+     * Field mappers must override this method if they provide
+     * a more efficient field-specific implementation of synthetic source.
+     * </p>
+     * @return {@link SyntheticSourceMode}
+     */
+    protected SyntheticSourceSupport syntheticSourceSupport() {
+        return SyntheticSourceSupport.FALLBACK;
+    }
+
+    /**
+     * Specifies the mode of synthetic source support by the mapper.
+     *
+     * <pre>
+     * {@link SyntheticSourceMode#NATIVE} - mapper natively supports synthetic source, f.e. by constructing it from doc values.
+     *
+     * {@link SyntheticSourceMode#FALLBACK} - mapper does not have native support and uses generic fallback implementation
+     * that stores raw input source data as is.
+     * </pre>
+     */
+    protected enum SyntheticSourceMode {
+        NATIVE,
+        FALLBACK
```

Comment:
Asked this above already: should we name these something like `WITH_DOC_VALUES` and `WITH_STORED_FIELDS`?